ocaml/xapi/xapi_host_driver.ml

```diff
@@ -85,6 +94,168 @@ let __reload_driver_selection ~__context ~host =
   with Forkhelpers.Spawn_internal_error (err, _, _) ->
     internal_error "Failed to reload driver selection (%s): %s" __FUNCTION__ err
 
+(* Versions are compared in segments separated by punctuation characters.
+   If both segments are integers, integer comparison is used, if at least
+   one of the segments is not an integer, lexicographic comparison is used.
+   Any segment compares larger against an empty segment.
+   Thus, "1.2.3" > "1.2";
+         "1.2.3" > "1.2."
+         "1.2.3" < "1.2.fix"
+         "1.2.3" < "1.2.3fix"
+*)
+module Version = struct
+  let compare a b =
+    let split str = Str.(split_delim (regexp "[.,:;+_-]") str) in
+
+    List.compare
+      (fun a b ->
+        Stdlib.(
+          match (int_of_string_opt a, int_of_string_opt b) with
+          | Some a, Some b ->
+              Int.compare a b
+          | _, _ ->
+              String.compare a b
+        )
+      )
+      (split a) (split b)
+end
+
+let __get_driver_versions driver_name selected_dir all_versions_dir =
```

Comment:
why `__`?

Reply:
Just a Python-style indication of an internal helper function given that `.mli` files do not play nice in this directory, I believe.

Reply:
ok, but I have never seen this kind of style being used on ocaml functions, maybe @lindig would know. What is wrong with the mli files?

Reply:
I've not seen the `_` prefixing - it suppresses the `unused function` warning, so would not use it to designate internal functions. But I'm a live-and-let-live guy.

Reply:
I've removed the underscores and created an `.mli` file